README.md
# is_pirated 🏴‍☠️
[![Pub Version](https://img.shields.io/pub/v/is_pirated)](https://pub.dev/packages/is_pirated/)

Check if your app is installed from an unauthorized source

On Android API 29 and below, uses `Context.getInstallerPackageName()`

On Android API 30 and above, uses `PackageManager.getInstallSourceInfo()`

On iOS, parses `Bundle.main.appStoreReceiptURL`


Installing
----------

Add to your pubspec.yaml

```yaml
dependencies:
  is_pirated: <version>
```


Example
-------
```dart
IsPirated isPirated = await getIsPirated();

print(isPirated.status);
// false - (iOS) if installed from AppStore or TestFlight, (Android) if installed from PlayStore
// true - if installed without authorization or receipt (Via ABD, APK, IPA)

//full example
IsPirated isPirated = await getIsPirated(debugOverride: true, openStoreListing: true, appStoreId: '546532666', playStoreIdentifier: 'tikmoji.sethusenthil.com');
//debugOverride will disable installer integrity checking on debug mode (recommended setting after you finish setting up this lib!)
//openStoreListing will auto open your appstore or playstore listing depending on the platform

```

## Options for .getIsPirated()

| Name                    | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                  |
| ------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| **debugOverride** | **bool** <br>Default set to false. Override unauthorized installation checking when on debug mode. Set this to true after configuring this library.                                                                                                                                                                                                                                                                                                                           |
| **openStoreListing** | **bool** <br>Default set to false. Open the AppStore or PlayStore listing if isPirated.status = true. Must set appStoreId & playStoreIdentifier for this to take any effect on its respective platform.                                                                                                                                                                                                                                                                                                                          |
| **appStoreId** | **String** <br>The AppStore App Id for your app so openStoreListing can open your store listing. You can find it in the url of your app: 'https://apps.apple.com/us/app/tikmoji/id**1546532666**' Must set openStoreListing true.                                                                                                                                                                                                                                                                                                                           |
| **playStoreIdentifier** | **String** <br>The PlayStore App Store Identifier for your app so it can open your listing on the PlayStore. You can find this in your manifest or your PlayStore URL: https://play.google.com/store/apps/details?id=**com.sethusenthil.tikmoji**                                                                                                                                                                                                                                                                                                                           |
| **closeApp** | **bool** <br>Default set to false. Closes and quits out of your app if isPirated.status = true. If openStoreListing is set to true, it will open your store listing before quitting. Note that getIsPirated() may not return a value if set to true because the Dart VM will likely be terminated before.                                                                                                                                                                                                                                                                                                                          |                                                                                                                                                                                                                                                                           |
## Credits
Inspired and heavely based on https://github.com/caiopo/flutter-installer-info


CHANGELOG.md
## 0.3.0
Updated URL Launcher dependencies for flutter 3.0
## 0.1.1
* Updated Docs and other Meta Data. No changes to actual library
## 0.1.0
* Initial release.

example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:is_pirated_example/main.dart';

void main() {
  testWidgets('Verify Platform version', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that platform version is retrieved.
    expect(
      find.byWidgetPredicate(
        (Widget widget) =>
            widget is Text && widget.data.startsWith('Running on:'),
      ),
      findsOneWidget,
    );
  });
}


test/installer_info_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:is_pirated/is_pirated.dart';

void main() {
  const MethodChannel channel = MethodChannel('com.sethusenthil.is_pirated');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      if (methodCall.method == 'getIsPirated') {
        return false;
      }
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getIsPirated', () async {
    final isPirated = await getIsPirated();
    expect(isPirated!.status is bool, true);
  });
}


example/lib/main.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:is_pirated/is_pirated.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  IsPirated isPirated;

  @override
  void initState() {
    super.initState();
    initPlatformState();
  }

  Future<void> initPlatformState() async {
    IsPirated isPirated;
    try {
      isPirated = await getIsPirated(debugOverride: false);

      //full example
      /*isPirated = await getIsPirated(
          debugOverride: true,
          openStoreListing: true,
          appStoreId: '546532666',
          playStoreIdentifier: 'tikmoji.sethusenthil.com',
          closeApp: true);*/
    } on PlatformException {}

    if (!mounted) return;

    setState(() => this.isPirated = isPirated);
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Is Pirated 🏴‍☠️'),
        ),
        body: Center(
          child: Text(
            'This app is ${isPirated?.status == false ? "NOT " : ""}pirated',
            textAlign: TextAlign.center,
            style: TextStyle(
                fontWeight: FontWeight.bold,
                color: isPirated?.status == true ? Colors.red : Colors.green),
          ),
        ),
      ),
    );
  }
}


ios/Classes/SwiftInstallerInfoPlugin.swift
import Flutter
import UIKit

public class SwiftIsPiratedPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "com.sethusenthil.is_pirated", binaryMessenger: registrar.messenger())
        let instance = SwiftIsPiratedPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }

    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if call.method == "getIsPirated" {
            if let path = Bundle.main.appStoreReceiptURL?.path {
                if path.contains("CoreSimulator") {
                    result("com.apple.CoreSimulator")
                } else if path.contains("sandboxReceipt") {
                    result("com.apple.TestFlight")
                } else {
                    result("com.apple.AppStore")
                }
            } else {
                result(nil)
            }
        } else {
            result(FlutterMethodNotImplemented)
        }
    }
}


android/src/main/kotlin/com/sethusenthil/is_pirated/IsPiratedPlugin.kt
package com.sethusenthil.is_pirated

import android.content.Context
import android.content.pm.InstallSourceInfo
import android.os.Build
import androidx.annotation.NonNull
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result


class IsPiratedPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var context: Context? = null

    override fun onAttachedToEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = binding.applicationContext
        channel = MethodChannel(binding.binaryMessenger, "com.sethusenthil.is_pirated")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        if (call.method == "getIsPirated") {
            val installerName = context?.run { getInstallerPackageName(this) }
            result.success(installerName)
        } else {
            result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        context = null
        channel.setMethodCallHandler(null)
    }

    private fun getInstallerPackageName(ctx: Context): String? {
        try {
            val packageName = ctx.packageName
            val pm = ctx.packageManager
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
                val info: InstallSourceInfo = pm.getInstallSourceInfo(packageName)
                return info.installingPackageName
            }
            return pm.getInstallerPackageName(packageName)
        } catch (e: Exception) {
            return null
        }
    }
}


lib/is_pirated.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/services.dart';
import 'package:flutter/foundation.dart' as Foundation;
import 'package:url_launcher/url_launcher.dart';

const _channel = MethodChannel('com.sethusenthil.is_pirated');

/// Returns if the app is installed authentically.
///
/// On Android API 29 and below, uses Context.getInstallerPackageName()
/// On Android API 30 and above, uses PackageManager.getInstallSourceInfo()
/// On iOS, parses Bundle.main.appStoreReceiptURL
Future<IsPirated?> getIsPirated({
  bool? debugOverride,
  bool? openStoreListing,
  bool? closeApp,
  String? appStoreId,
  String? playStoreIdentifier,
}) async {
  final installerName = await _channel.invokeMethod('getIsPirated');

  bool isItPirated = false;
  //Return True if installed WITHOUT proper store autherositaion
  //Retuns False if installed through proper app store

  debugOverride ??= false;
  openStoreListing ??= false;
  closeApp ??= false;

// Installers whitelist
  const installers = [
    // Android
    'com.android.vending',

    // iOS
    'com.apple.AppStore',
    'com.apple.TestFlight',
    'com.apple.CoreSimulator',
  ];

  bool validInstaller = false;

  if (installerName != null) {
    installers.forEach((installer) {
      if (installerName == installer) {
        validInstaller = true;
      }
    });
  }

  if (!validInstaller && (debugOverride ? Foundation.kReleaseMode : true)) {
    isItPirated = true;

    if (openStoreListing) {
      if (Platform.isIOS && appStoreId != null) {
        if (appStoreId.startsWith("id")) appStoreId = appStoreId.split("id")[1];
        launchUrl(Uri.parse('https://apps.apple.com/app/id' + appStoreId),
            mode: LaunchMode.externalApplication);
      }
      if (Platform.isAndroid && playStoreIdentifier != null)
        launchUrl(
            Uri.parse("https://play.google.com/store/apps/details?id=" +
                playStoreIdentifier),
            mode: LaunchMode.externalApplication);
    }

    if (closeApp) exit(0);
  }

  return IsPirated(isItPirated);
}

class IsPirated {
  /// The installer package name, as returned by the platform
  final bool status;

  /// Bool value based on [installerName]

  IsPirated(this.status);
}
